src/data/repositories/MasterLogFrameD2Repository.ts

```diff
@@ -0,0 +1,169 @@
+                    fields: indicatorGroupFields,
+                    filter: { id: { in: ids } },
+                })
+            ).map(response =>
+                response.objects.map(d2Group => this.buildMasterLogFrame(d2Group, "outcomes"))
+            );
+        });
+    }
+
+    private buildMasterLogFrame(
+        d2Group: D2DataElementGroup | D2IndicatorGroup,
+        indicatorType: IndicatorType
+    ): MasterLogFrame {
+        const items =
+            indicatorType === "outputs" && this.isDataElementGroup(d2Group)
+                ? d2Group.dataElements
+                : indicatorType === "outcomes" && this.isIndicatorGroup(d2Group)
+                ? d2Group.indicators
+                : [];
+
+        const description = this.isDataElementGroup(d2Group)
+            ? d2Group.displayDescription
+            : d2Group.description;
+
+        return {
+            id: d2Group.id,
+            name: description || d2Group.displayName,
+            type: indicatorType,
+            indicators: _(items)
+                .map(d2DataElement => ({ id: d2DataElement.id }))
+                .sortBy(indicator => indicator.id)
+                .value(),
+        };
+    }
+
+    private getDataElementGroupSet(codes: string[]): FutureData<Id[]> {
+        return apiToFuture(
+            this.api.models.dataElementGroupSets.get({
+                fields: { dataElementGroups: { id: true, code: true, name: true } },
+                paging: false,
+                filter: { code: { eq: this.MLF_CODE } },
+            })
+        ).map(response => {
+            const d2GroupSet = response.objects[0];
+            return d2GroupSet
+                ? this.filterByCodeAndExtractIds(codes, d2GroupSet.dataElementGroups)
+                : [];
+        });
+    }
+
+    private getIndicatorGroupSet(codes: string[]): FutureData<Id[]> {
+        return apiToFuture(
+            this.api.models.indicatorGroupSets.get({
+                fields: { id: true, indicatorGroups: { id: true, code: true, name: true } },
+                paging: false,
+                filter: { code: { eq: this.MLF_CODE } },
+            })
+        ).map(response => {
+            const d2GroupSet = response.objects[0];
+            return d2GroupSet
+                ? this.filterByCodeAndExtractIds(codes, d2GroupSet.indicatorGroups)
+                : [];
+        });
+    }
+
+    private filterByCodeAndExtractIds(
+        codes: string[],
+        groups: Array<{ id: Id; code: string; name: string }>
+    ): Id[] {
+        return groups
+            .filter(group => codes.includes(extractRegionCode(group.code ?? group.name)))
+            .map(group => group.id);
+    }
+
+    private isDataElementGroup(
+        group: D2DataElementGroup | D2IndicatorGroup
+    ): group is D2DataElementGroup {
+        return "dataElements" in group;
+    }
+
+    private isIndicatorGroup(
+        group: D2DataElementGroup | D2IndicatorGroup
+    ): group is D2IndicatorGroup {
```

Comment:
As a rule of thumb, we try to avoid type guards in code under our control. I guess the necessity for this is having `d2Group: D2DataElementGroup | D2IndicatorGroup`. A typical approach is adding a field "type" to these two types, and now we have 100% type-safe discrimination.